File: c/src/mac.c
#include "mac.h"

#include <openssl/crypto.h>
#include <openssl/hmac.h>

#include "dorado/blake3.h"
#include "dorado/engine.h"
#include "dorado/skein.h"

void dorado_mac_tag(int mac, const uint8_t mac_key[32], const uint8_t *signed_data, size_t signed_len,
                    uint8_t out[32]) {
    switch (mac) {
        case DORADO_MAC_SKEIN:
            dorado_skein512_mac(mac_key, 32, 32, signed_data, signed_len, out);
            break;
        case DORADO_MAC_BLAKE3:
            dorado_blake3_keyed_mac(mac_key, 32, signed_data, signed_len, out);
            break;
        case DORADO_MAC_HMAC: {
            unsigned int len = 32;
            HMAC(EVP_sha256(), mac_key, 32, signed_data, signed_len, out, &len);
            break;
        }
        default:
            break;
    }
}

int dorado_mac_verify(int mac, const uint8_t mac_key[32], const uint8_t *signed_data, size_t signed_len,
                      const uint8_t received[32]) {
    uint8_t expected[32];
    dorado_mac_tag(mac, mac_key, signed_data, signed_len, expected);
    return CRYPTO_memcmp(expected, received, 32) == 0;
}

File: zig/src/mac.zig
//! The MAC menu: Skein-512 (default) and keyed BLAKE3 from this project's
//! from-scratch hashes, HMAC-SHA256 from Zig's standard library. All three take the
//! 32-byte MAC key and produce a 32-byte tag.

const std = @import("std");
const HmacSha256 = std.crypto.auth.hmac.sha2.HmacSha256;
const fmt = @import("format.zig");
const skein = @import("skein.zig");
const blake3 = @import("blake3.zig");

/// Compute a 32-byte tag over signed_data under the selected MAC.
pub fn tag(mac: fmt.Mac, mac_key: *const [32]u8, signed_data: []const u8, out: *[32]u8) void {
    switch (mac) {
        .skein => skein.mac(mac_key, 32, signed_data, out),
        .blake3 => blake3.keyedMac(mac_key, signed_data, out),
        .hmac => HmacSha256.create(out, signed_data, mac_key),
    }
}

/// Recompute the tag and compare it in constant time.
pub fn verify(mac: fmt.Mac, mac_key: *const [32]u8, signed_data: []const u8, received: *const [32]u8) bool {
    var expected: [32]u8 = undefined;
    tag(mac, mac_key, signed_data, &expected);
    return std.crypto.timing_safe.eql([32]u8, expected, received.*);
}
